test/remote/gateways/remote_shift4_test.rb

```diff
@@ -0,0 +1,247 @@
+  end
+
+  def test_failed_refund
+    response = @gateway.refund(nil, @credit_card, @refund_options.except(:access_token))
+
+    assert_failure response
+    assert gen_error_msg(response)
+  end
+
+  def test_successful_refund
+    res = @gateway.purchase(@amount, @token, @purchase_params)
+    transaction = res.params['result'][0]['transaction']
+    card = res.params['result'][0]['card']
+    invoice = transaction['invoice']
+    auth_token = card['token']['value']
+    @refund_options[:invoice] = invoice
+    assert_success res
+
+    response = @gateway.refund(@amount, auth_token, @refund_options)
+
+    assert_success response
+    assert_equal @refund_options[:total], verify_amount(response)
+    assert_equal @refund_options[:invoice], response_result(response)['transaction']['invoice']
+  end
+
+  def test_successful_void
+    authorize_res = @gateway.authorize(@amount, @credit_card, @options)
+    invoice = authorize_res.params['result'][0]['transaction']['invoice']
+    @void_options[:invoice] = invoice
+    assert response = @gateway.void(@void_options)
+
+    assert_success response
+    assert_equal @refund_options[:total], verify_amount(response)
+    assert_equal @void_options[:invoice], response_result(response)['transaction']['invoice']
+  end
+
+  def test_failed_void
+    response = @gateway.void(@void_options.except(:invoice))
+    assert_failure response
+    assert void_error_msg(response)
+    assert_equal error_code(response)['primaryCode'], 9815
+  end
+
+  private
+
+  def access_token_success_from(response)
+    credentials(response)['accessToken'].present?
+  end
+
+  def credentials(response)
+    response_result(response)['credential']
+  end
+
+  def successful_transaction?(response)
+    token(response).present?
+  end
+
+  def token(response)
+    response_result(response)['card']['token']['value']
+  end
+
+  def success?(response)
+    response_result(response)['error'].nil?
+  end
+
+  def response_result(response)
+    response.params['result'][0]
+  end
+
+  def success_from(response)
+    response_result(response)['credential']['accessToken'].present?
+  end
+
+  def verify_amount(response)
+    response_result(response)['amount']['total']
+  end
+
+  def product_des(response)
+    response_result(response)['transaction']['purchaseCard']['productDescriptors'].first
+  end
+
+  def error_msg(response)
+    error_text(response).include?('No GTV PAN returned from host')
+  end
+
+  def merchant_error(response)
+    error_text(response).include?('GTV Msg: ERROR{0} 20018: no default category found, UC')
+  end
+
+  def declined_card_error(response)
+    error_text(response)
+  end
+
+  def gen_error_msg(response)
+    error_text(response).include?('Secondary amounts cannot exceed the total amount')
+  end
+
+  def void_error_msg(response)
+    error_text(response).include?('Invoice Not Found')
+  end
+
+  def error_text(response)
+    error(response)['longText']
+  end
+
+  def error(response)
```

Comment:
You should remove these and just retrieve the necessary values in the same line you make the asserts. What you end up with here is a bunch of nested method calls that can be difficult to triage if we run into errors and it also looks cluttered.